R/trace_tests.R

```diff
@@ -97,17 +97,60 @@ NULL
 count_test <- function(key) {
   n_calls_into_covr <- 2L
 
-  if (is_current_test_finished())
-    update_current_test(key)
+  if (is_current_test_finished()) {
+    update_current_test()
+  }
 
   # ignore if .counter was not created with record_tests (nested coverage calls)
   if (is.null(.counters[[key]]$tests)) return()
 
-  depth_into_pkg <- length(sys.calls()) - length(.current_test$frames) - n_calls_into_covr + 1L
   .current_test$i <- .current_test$i + 1L
-  .counters[[key]]$tests <- rbind(
-    .counters[[key]]$tests,
-    c(length(.counters$tests), depth_into_pkg, .current_test$i)
+
+  # expand infrequently as new tests are added, doubling matrix size as needed
+  tests <- .counters[[key]]$tests
+  n <- NROW(tests$tally)
+  if (.counters[[key]]$value > n) {
+    tests$tally <- rbind(tests$tally, matrix(NA_integer_, ncol = 3L, nrow = n))
+  }
+
+  # test number
+  tests$.data[[1L]] <- length(.counters$tests)
+
+  # call stack depth when trace is hit
+  tests$.data[[2L]] <- sys.nframe() - length(.current_test$frames) - n_calls_into_covr + 1L
+
+  # number of traces hit by the test so far
+  tests$.data[[3L]] <- .current_test$i
+
+  tests$.value <- .counters[[key]]$value
+  with(tests, tally[.value,] <- .data)
```

Comment:
The nested indexed assignment (`.counters[[key]]$tests <- ...`) was also resulting in a big performance hit due to reallocations. 

Instead, `.counters[[key]]$tests` is now an `environment` so that `.counters[[key]]$tests$tally` can be modified in place and avoid drilling down into the counters object.